lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,207 @@
+//===- DpiEntryPoints.cpp - ESI cosim DPI calls -----------------*- C++ -*-===//
+//
+// Cosim DPI function implementations. Mostly C-C++ gaskets to the C++
+// RpcServer.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Server.h"
+#include "circt/Dialect/ESI/cosim/dpi.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+
+using namespace circt::esi::cosim;
+
+static RpcServer *server = nullptr;
+
+// ---- Helper functions ----
+
+/// Get the TCP port on which to listen. Defaults to 0xECD (ESI Cosim DPI), 3789
+/// in decimal.
+int getPort() {
+  const char *portEnv = getenv("COSIM_PORT");
+  if (portEnv == nullptr) {
+    printf("[COSIM] RPC server port not found. Defaulting to 3789.\n");
+    return 0xECD;
+  }
+  printf("[COSIM] Opening RPC server on port %s\n", portEnv);
+  return std::strtoull(portEnv, nullptr, 10);
+}
+
+/// Check that an array is an array of bytes and has some size.
+// NOLINTNEXTLINE(misc-misplaced-const)
+static int validateSvOpenArray(const svOpenArrayHandle data,
+                               int expectedElemSize) {
+  if (svDimensions(data) != 1) {
+    printf("DPI-C: ERROR passed array argument that doesn't have expected 1D "
+           "dimensions\n");
+    return -1;
+  }
+  if (svGetArrayPtr(data) == NULL) {
+    printf("DPI-C: ERROR passed array argument that doesn't have C layout "
+           "(ptr==NULL)\n");
+    return -2;
+  }
+  int totalBytes = svSizeOfArray(data);
+  if (totalBytes == 0) {
+    printf("DPI-C: ERROR passed array argument that doesn't have C layout "
+           "(total_bytes==0)\n");
+    return -3;
+  }
+  int numElems = svSize(data, 1);
+  int elemSize = numElems == 0 ? 0 : (totalBytes / numElems);
+  if (numElems * expectedElemSize != totalBytes) {
+    printf("DPI-C: ERROR: passed array argument that doesn't have expected "
+           "element-size: expected=%d actual=%d\n",
+           expectedElemSize, elemSize);
+    return -4;
+  }
+  return 0;
+}
+
+// ---- DPI entry points ----
+
+// Register simulated device endpoints.
+// - return 0 on success, non-zero on failure (duplicate EP registered).
+DPI int sv2cCosimserverEpRegister(int endpointId, long long sendTypeId,
+                                  int sendTypeSize, long long recvTypeId,
```

Comment:
Please use int64_t instead of long long.  llvm/Support/DataTypes.h provides it.

Reply:
`long long` is not a type I would use. It's generated by the RTL simulator (from `CosimDpiPkg.sv`) and called by the RTL simulator so I'm hesitant to make even a change which wouldn't break anything.